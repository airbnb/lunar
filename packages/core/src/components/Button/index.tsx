import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import ButtonOrLink, { Props as ButtonOrLinkProps } from '../private/ButtonOrLink';
import Loader from '../Loader';

const sizingProp = mutuallyExclusiveTrueProps('small', 'large');

const GOLDEN_RATIO = (1 + Math.sqrt(5)) / 2;

export type Props = ButtonOrLinkProps & {
  /** Render as a block with full width. */
  block?: boolean;
  /** Render as borderless. */
  borderless?: boolean;
  /** @ignore Hidden prop used in forms. */
  invalid?: boolean;
  /** Invert text colors. */
  inverted?: boolean;
  /** Increase font size to large. */
  large?: boolean;
  /** Decrease font size to small. */
  small?: boolean;
};

/** A standard button and or link for common UI interactions. */
export class Button extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    large: sizingProp,
    small: sizingProp,
  };

  static defaultProps = {
    block: false,
    borderless: false,
    disabled: false,
    invalid: false,
    inverted: false,
    large: false,
    loading: false,
    small: false,
  };

  render() {
    const {
      cx,
      block,
      borderless,
      children,
      disabled,
      invalid,
      inverted,
      large,
      loading,
      small,
      styles,
      ...restProps
    } = this.props;

    return (
      <ButtonOrLink
        {...restProps}
        aria-busy={loading}
        disabled={disabled}
        loading={loading}
        className={cx(
          styles.button,
          large && styles.button_large,
          small && styles.button_small,
          !large && !small && styles.button_regular,
          block && styles.button_block,
          (disabled || loading) && styles.button_disabled,
          (borderless || inverted) && styles.button_inverted,
          invalid && styles.button_invalid,
          borderless && styles.button_borderless,
          loading && styles.button_loading,
        )}
      >
        {loading ? <Loader inline inverted={!inverted} /> : children}
      </ButtonOrLink>
    );
  }
}

export default withStyles(
  ({ color, font, pattern, ui, unit, transition }) => ({
    button: {
      ...pattern.resetButton,
      ...transition.box,
      fontWeight: font.weights.bold,
      position: 'relative',
      color: color.base,
      backgroundColor: color.core.primary[3],
      border: `2px solid ${color.core.primary[3]}`,
      borderRadius: ui.borderRadius,
      textAlign: 'center',

      '@selectors': {
        // Removes weird bonus padding from button in Firefox
        '::-moz-focus-inner': {
          border: 0,
          padding: 0,
          margin: 0,
        },

        // Only show hover states on non-disabled
        ':not([disabled]):hover': {
          backgroundColor: color.core.primary[4],
          borderColor: color.core.primary[4],
        },
      },
    },

    button_block: {
      display: 'block',
      width: '100%',
      whiteSpace: 'normal',
      overflow: 'hidden',
    },

    button_borderless: {
      borderColor: 'transparent',

      '@selectors': {
        ':not([disabled]):hover': {
          borderColor: color.accent.bgHover,
        },
      },
    },

    button_disabled: {
      ...pattern.disabled,
    },

    button_invalid: {},

    button_inverted: {
      color: color.core.primary[3],
      backgroundColor: color.accent.bg,

      '@selectors': {
        ':not([disabled]):hover': {
          color: color.core.primary[4],
          backgroundColor: color.accent.bgHover,
        },
      },
    },

    button_loading: {
      cursor: 'default',
    },

    button_small: {
      ...pattern.smallButton,
      minWidth: GOLDEN_RATIO * 4 * unit,
    },

    button_regular: {
      ...pattern.regularButton,
      minWidth: GOLDEN_RATIO * 6 * unit,
    },

    button_large: {
      ...pattern.largeButton,
      minWidth: GOLDEN_RATIO * 8 * unit,
    },
  }),
  {
    extendable: true,
  },
)(Button);
