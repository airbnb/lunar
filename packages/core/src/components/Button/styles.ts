import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, font, pattern, ui, unit, transition }) => ({
  button: {
    ...pattern.resetButton,
    ...transition.box,
    fontWeight: font.weights.bold,
    position: 'relative',
    color: color.base,
    backgroundColor: color.core.primary[3],
    border: `${ui.borderWidthThick}px solid ${color.core.primary[3]}`,
    borderRadius: ui.borderRadius,
    textAlign: 'center',

    '@selectors': {
      // Removes weird bonus padding from button in Firefox
      '::-moz-focus-inner': {
        border: 0,
        padding: 0,
        margin: 0,
      },

      // Only show hover states on non-disabled
      ':not([disabled]):hover': {
        backgroundColor: color.core.primary[4],
        borderColor: color.core.primary[4],
      },
    },
  },

  button_block: {
    display: 'block',
    width: '100%',
    whiteSpace: 'normal',
    overflow: 'hidden',
  },

  button_borderless: {
    borderColor: 'transparent',

    '@selectors': {
      ':not([disabled]):hover': {
        borderColor: color.accent.bgHover,
      },
    },
  },

  button_disabled: {
    ...pattern.disabled,
  },

  button_invalid: {},

  button_inverted: {
    color: color.core.primary[3],
    backgroundColor: color.accent.bg,

    '@selectors': {
      ':not([disabled]):hover': {
        color: color.core.primary[4],
        backgroundColor: color.accent.bgHover,
      },
    },
  },

  button_loading: {
    cursor: 'default',
  },

  button_small: {
    ...pattern.smallButton,
    minWidth: 6 * unit,
  },

  button_regular: {
    ...pattern.regularButton,
    minWidth: 8 * unit,
  },

  button_large: {
    ...pattern.largeButton,
    minWidth: 9 * unit,
  },
});

export default styleSheet;

export { styleSheet };
