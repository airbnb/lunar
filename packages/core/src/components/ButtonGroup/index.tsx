import React from 'react';
import childrenWithComponentName from '../../prop-types/childrenWithComponentName';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** List of components to group. */
  children: NonNullable<React.ReactNode>;
  /** Stack the buttons vertically. */
  stacked?: boolean;
};

/** Horizontally align `Button`s with a consistent gutter between each. */
export class ButtonGroup extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    children: childrenWithComponentName('Buttons?|Links?').isRequired,
  };

  static defaultProps = {
    stacked: false,
  };

  render() {
    const { children, stacked, styles } = this.props;

    return (
      <div {...css(styles.buttonGroup, stacked && styles.buttonGroup_stacked)}>
        {React.Children.map(children, child =>
          child ? <div {...css(stacked ? styles.cell_stacked : styles.cell)}>{child}</div> : null,
        )}
      </div>
    );
  }
}

export default withStyles(({ unit }) => ({
  buttonGroup: {
    display: 'flex',
    alignItems: 'center',
  },

  buttonGroup_stacked: {
    flexDirection: 'column',
    alignItems: 'stretch',
  },

  cell: {
    marginRight: unit,

    ':last-of-type': {
      marginRight: 0,
    },

    ':empty': {
      display: 'none',
    },
  },

  cell_stacked: {
    marginBottom: unit,

    ':last-of-type': {
      marginBottom: 0,
    },
  },
}))(ButtonGroup);
