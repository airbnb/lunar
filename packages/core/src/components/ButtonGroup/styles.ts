import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ unit }) => ({
  buttonGroup: {
    display: 'flex',
    alignItems: 'center',
  },

  buttonGroup_stacked: {
    flexDirection: 'column',
    alignItems: 'stretch',
  },

  cell: {
    marginRight: unit,

    ':last-of-type': {
      marginRight: 0,
    },

    ':empty': {
      display: 'none',
    },
  },

  cell_stacked: {
    marginBottom: unit,

    ':last-of-type': {
      marginBottom: 0,
    },
  },
});

export default styleSheet;

export { styleSheet };
