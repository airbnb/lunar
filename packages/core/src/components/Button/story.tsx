import React from 'react';
import IconAddAlt from '@airbnb/lunar-icons/lib/interface/IconAddAlt';
import Button from '.';

export default {
  title: 'Core/Button',
  parameters: {
    inspectComponents: [Button],
  },
};

export function standardButtonWithAnInvertedVariant() {
  return (
    <>
      <Button>Button</Button> <Button inverted>Button</Button>
    </>
  );
}

standardButtonWithAnInvertedVariant.story = {
  name: 'Standard button with an inverted variant.',
};

export function renderAnAnchorLinkWhenPassingHref() {
  return (
    <>
      <Button openInNewWindow href="https://github.com/airbnb/lunar">
        Link
      </Button>{' '}
      <Button openInNewWindow inverted href="https://github.com/airbnb/lunar">
        Link
      </Button>
    </>
  );
}

renderAnAnchorLinkWhenPassingHref.story = {
  name: 'Render an anchor link when passing `href`.',
};

export function withEventHandlers() {
  return (
    <Button
      onClick={action('onClick')}
      onMouseOver={action('onMouseOver')}
      onFocus={action('onFocus')}
    >
      Button
    </Button>
  );
}

withEventHandlers.story = {
  name: 'With event handlers.',
};

export function withDifferentSizingSmallRegularDefaultAndLarge() {
  return (
    <>
      <Button small>Button</Button> <Button>Button</Button> <Button large>Button</Button>
      <br />
      <br />
      <Button inverted small>
        Button
      </Button>{' '}
      <Button inverted>Button</Button>{' '}
      <Button inverted large>
        Button
      </Button>
    </>
  );
}

withDifferentSizingSmallRegularDefaultAndLarge.story = {
  name: 'With different sizing: small, regular (default), and large.',
};

export function withDifferentStatesDisabledAndLoading() {
  return (
    <>
      <Button disabled>Button</Button> <Button loading>Button</Button>
      <br />
      <br />
      <Button inverted disabled>
        Button
      </Button>{' '}
      <Button inverted loading>
        Button
      </Button>
    </>
  );
}

withDifferentStatesDisabledAndLoading.story = {
  name: 'With different states: disabled and loading.',
};

export function withBeforeAndOrAfterIcons() {
  return (
    <>
      <div>
        <Button beforeIcon={<IconAddAlt decorative />}>Before icon</Button>{' '}
        <Button afterIcon={<IconAddAlt decorative />}>After icon</Button>
      </div>
    </>
  );
}

withBeforeAndOrAfterIcons.story = {
  name: 'With before and or after icons.',
};

export function withBorderless() {
  return (
    <>
      <div>
        <Button borderless>Button</Button>{' '}
        <Button borderless inverted>
          Button
        </Button>{' '}
        <Button borderless disabled>
          Button
        </Button>
      </div>
    </>
  );
}

withBorderless.story = {
  name: 'With borderless.',
};
