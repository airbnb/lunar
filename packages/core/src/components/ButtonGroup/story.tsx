import React from 'react';
import Button from '../Button';
import Tooltip from '../Tooltip';
import ButtonGroup from '.';

export default {
  title: 'Core/ButtonGroup',
  parameters: {
    inspectComponents: [ButtonGroup],
  },
};

export function groupAnArbitraryNumberOfComponents() {
  return (
    <ButtonGroup>
      <Button>One</Button>
      <Button>Two</Button>
      <Button>Three</Button>
    </ButtonGroup>
  );
}

groupAnArbitraryNumberOfComponents.story = {
  name: 'Group an arbitrary number of components.',
};

export function stackButtonsVertically() {
  return (
    <div style={{ width: 300 }}>
      <ButtonGroup stacked>
        <Button block small>
          One
        </Button>
        <Button block>Two</Button>
        <Button block large>
          Three
        </Button>
      </ButtonGroup>
    </div>
  );
}

stackButtonsVertically.story = {
  name: 'Stack buttons vertically.',
};

export function buttonsWrappedInATooltip() {
  return (
    <ButtonGroup>
      <Tooltip content="One">
        <Button>One</Button>
      </Tooltip>
      <Tooltip content="Two">
        <Button>Two</Button>
      </Tooltip>
      <Tooltip content="Three">
        <Button>Three</Button>
      </Tooltip>
    </ButtonGroup>
  );
}

buttonsWrappedInATooltip.story = {
  name: 'Buttons wrapped in a `Tooltip`.',
};
