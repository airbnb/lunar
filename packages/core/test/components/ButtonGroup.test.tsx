import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import ButtonGroup from '../../src/components/ButtonGroup';
import Button from '../../src/components/Button';

describe('<ButtonGroup />', () => {
  it('renders buttons', () => {
    const wrapper = shallowWithStyles(
      <ButtonGroup>
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </ButtonGroup>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders buttons stacked', () => {
    const wrapper = shallowWithStyles(
      <ButtonGroup stacked>
        <Button>One</Button>
        <Button>Two</Button>
        <Button>Three</Button>
      </ButtonGroup>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a single button', () => {
    const wrapper = shallowWithStyles(
      <ButtonGroup>
        <Button>One</Button>
      </ButtonGroup>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('handles buttons that return falsy values', () => {
    const wrapper = shallowWithStyles(
      <ButtonGroup stacked>
        {false && <Button>One</Button>}
        {true && <Button>Two</Button>}
        {null && <Button>Three</Button>}
      </ButtonGroup>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('handles components that return a falsy value', () => {
    function FakeButton() {
      return null;
    }

    const wrapper = shallowWithStyles(
      <ButtonGroup stacked>
        <FakeButton />
      </ButtonGroup>,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
