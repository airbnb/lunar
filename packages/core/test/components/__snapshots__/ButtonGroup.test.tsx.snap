// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> errors when a non-button is passed 1`] = `
"Warning: Failed prop type: \`ButtonGroup.children\` only accepts components matching the regular expression /(?:Tooltip|Permission|Trebuchet|ERF)|(?:Buttons?|Links?)$/
    in ButtonGroup"
`;

exports[`<ButtonGroup /> handles buttons that return falsy values 1`] = `
<div
  className="buttonGroup_5kaapu-o_O-buttonGroup_stacked_ohgw5r"
>
  <div
    className="cell_stacked_123jc7u"
    key=".1"
  >
    <Button>
      Two
    </Button>
  </div>
</div>
`;

exports[`<ButtonGroup /> handles components that return a falsy value 1`] = `
<div
  className="buttonGroup_5kaapu-o_O-buttonGroup_stacked_ohgw5r"
>
  <div
    className="cell_stacked_123jc7u"
    key=".0"
  >
    <FakeButton />
  </div>
</div>
`;

exports[`<ButtonGroup /> renders a single button 1`] = `
<div
  className="buttonGroup_5kaapu"
>
  <div
    className="cell_1cks6un"
    key=".0"
  >
    <Button>
      One
    </Button>
  </div>
</div>
`;

exports[`<ButtonGroup /> renders buttons 1`] = `
<div
  className="buttonGroup_5kaapu"
>
  <div
    className="cell_1cks6un"
    key=".0"
  >
    <Button>
      One
    </Button>
  </div>
  <div
    className="cell_1cks6un"
    key=".1"
  >
    <Button>
      Two
    </Button>
  </div>
  <div
    className="cell_1cks6un"
    key=".2"
  >
    <Button>
      Three
    </Button>
  </div>
</div>
`;

exports[`<ButtonGroup /> renders buttons stacked 1`] = `
<div
  className="buttonGroup_5kaapu-o_O-buttonGroup_stacked_ohgw5r"
>
  <div
    className="cell_stacked_123jc7u"
    key=".0"
  >
    <Button>
      One
    </Button>
  </div>
  <div
    className="cell_stacked_123jc7u"
    key=".1"
  >
    <Button>
      Two
    </Button>
  </div>
  <div
    className="cell_stacked_123jc7u"
    key=".2"
  >
    <Button>
      Three
    </Button>
  </div>
</div>
`;
